### PR TITLE
[Feature] Add box_iou_quadri & nms_quadri

### PR DESCRIPTION
**Pull request by zytx121**
## Motivation

Add `box_iou_quadri` and `nms_quadri` ops, which used by quadrilateral boxes detection in https://github.com/open-mmlab/mmrotate/pull/520

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcv/ops/csrc/common/box_iou_rotated_utils.hpp [hunk 1]
```diff
@@ -270,6 +270,16 @@ HOST_DEVICE_INLINE int convex_hull_graham(const Point<T> (&p)[24],
   return m;
 }
 
+template <typename T>
+HOST_DEVICE_INLINE T quadri_box_area(const Point<T> (&q)[4]) {
+  T area = 0;
+  for (int i = 1; i < 3; i++) {
```
Comment (grimoire) [inline mmcv/ops/csrc/common/box_iou_rotated_utils.hpp:276]: This loop can be unrolled:

```c++
#pragma unroll
for(...)
```

File: mmcv/ops/csrc/pytorch/nms_quadri.cpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+#include "pytorch_cpp_helper.hpp"
+
+Tensor nms_quadri_cpu(const Tensor dets, const Tensor scores,
```
Comment (grimoire) [inline mmcv/ops/csrc/pytorch/nms_quadri.cpp:4]: Could we register the implementation in `cpu/nms_quadri.cpp` and `cuda/cudabind.cpp`?

Comment (zytx121) [reply mmcv/ops/csrc/pytorch/nms_quadri.cpp:4]: The input parameters of cpu and cuda is different (mmcv/ops/csrc/pytorch/nms_quadri.cpp). So we cannot register it as same input.

File: mmcv/ops/box_iou_quadri.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    of bboxes1 and bboxes2, otherwise the ious between each aligned pair of
+    bboxes1 and bboxes2.
+
+    Args:
+        boxes1 (torch.Tensor): quadrilateral bboxes 1. It has shape (N, 8),
+            indicating (x1, y1, ..., x4, y4) for each row. Note that theta is
+            in radian.
+        boxes2 (torch.Tensor): quadrilateral bboxes 2. It has shape (M, 8),
+            indicating (x1, y1, ..., x4, y4) for each row. Note that theta is
+            in radian.
```
Comment (zhouzaida) [inline mmcv/ops/box_iou_quadri.py:28]: Is there theta in `boxes`?

File: mmcv/ops/box_iou_quadri.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    Both sets of boxes are expected to be in
+    (x1, y1, ..., x4, y4) format.
+
+    If ``aligned`` is ``False``, then calculate the ious between each bbox
+    of bboxes1 and bboxes2, otherwise the ious between each aligned pair of
+    bboxes1 and bboxes2.
+
+    Args:
+        boxes1 (torch.Tensor): quadrilateral bboxes 1. It has shape (N, 8),
```
Comment (zhouzaida) [inline mmcv/ops/box_iou_quadri.py:23]:
```suggestion
        bboxes1 (torch.Tensor): quadrilateral bboxes 1. It has shape (N, 8),
```

File: mmcv/ops/box_iou_quadri.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            indicating (x1, y1, ..., x4, y4) for each row. Note that theta is
+            in radian.
+        boxes2 (torch.Tensor): quadrilateral bboxes 2. It has shape (M, 8),
```
Comment (zhouzaida) [inline mmcv/ops/box_iou_quadri.py:26]:
```suggestion
        bboxes2 (torch.Tensor): quadrilateral bboxes 2. It has shape (M, 8),
```

File: mmcv/ops/nms.py [hunk 6]
```diff
@@ -459,3 +459,44 @@ def nms_rotated(dets: Tensor,
...
+        same data type as the input.
+    """
+    if dets.shape[0] == 0:
+        return dets, None
+
+    multi_label = labels is not None
+    if multi_label:
+        dets_wl = torch.cat((dets, labels.unsqueeze(1)), 1)  # type: ignore
+    else:
+        dets_wl = dets
```
Comment (zhouzaida) [inline mmcv/ops/nms.py:494]:
```suggestion
        dets_with_lables = torch.cat((dets, labels.unsqueeze(1)), 1)  # type: ignore
    else:
        dets_with_lables = dets
```

File: tests/test_ops/test_nms_quadri.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        np_expect_keep_inds = np.array([3, 1, 2], dtype=np.int64)
+
+        boxes = torch.from_numpy(np_boxes)
+
+        dets, keep_inds = nms_quadri(boxes[:, :8], boxes[:, -1], 0.3)
+        assert np.allclose(dets.numpy()[:, :8], np_expect_dets)
+        assert np.allclose(keep_inds.numpy(), np_expect_keep_inds)
+
+    @pytest.mark.skipif(
+        not torch.cuda.is_available(), reason='requires CUDA support')
```
Comment (zhouzaida) [inline tests/test_ops/test_nms_quadri.py:78]: Same as above

File: tests/test_ops/test_nms_quadri.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+            torch.from_numpy(np_boxes[:, :8]),
+            torch.from_numpy(np_boxes[:, -1]),
+            torch.from_numpy(np_labels),
+            nms_cfg,
+            class_agnostic=False)
+        assert np.allclose(boxes.numpy()[:, :8], np_expect_dets)
+        assert np.allclose(keep.numpy(), np_expect_keep_inds)
+
+    @pytest.mark.skipif(
+        not torch.cuda.is_available(), reason='requires CUDA support')
```
Comment (zhouzaida) [inline tests/test_ops/test_nms_quadri.py:146]: same as above

File: tests/test_ops/test_nms_quadri.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+        assert np.allclose(boxes.numpy()[:, :8], np_expect_dets)
+        assert np.allclose(keep.numpy(), np_expect_keep_inds)
+
+    @pytest.mark.parametrize('device', [
+        pytest.param(
+            'cuda',
+            marks=pytest.mark.skipif(
+                not IS_CUDA_AVAILABLE, reason='requires CUDA support')),
+    ])
+    def test_batched_nms(self):
```
Comment (zhouzaida) [inline tests/test_ops/test_nms_quadri.py:161]:
```suggestion
    def test_batched_nms(self, device):
```

File: tests/test_ops/test_nms_quadri.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+                                   [7., 7., 8., 8., 9., 7., 8., 6.]],
+                                  dtype=np.float32)
+        np_expect_keep_inds = np.array([3, 1, 0, 2], dtype=np.int64)
+
+        nms_cfg = dict(type='nms_quadri', iou_threshold=0.3)
+
+        # test class_agnostic is True
+        boxes, keep = batched_nms(
+            torch.from_numpy(np_boxes[:, :8]).cuda(),
+            torch.from_numpy(np_boxes[:, -1]).cuda(),
```
Comment (zhouzaida) [inline tests/test_ops/test_nms_quadri.py:190]:
```suggestion
            torch.from_numpy(np_boxes[:, :8]).to(device),
            torch.from_numpy(np_boxes[:, -1]).to(device),
```

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**